File: web/src/queries/buildDetect.ts
// Fetcher for POST /api/v1/build/detect
// (internal/api/build_detect.go's handleDetectFramework): a fast,
// build-free pre-flight check of what Railpack would detect for a git
// source, for CreateAppFromGitFields.tsx to show "Detected: Next.js"
// before the operator commits to a build type. Not a TanStack Query
// `useQuery` hook, unlike queries/gitBranches.ts's useGitBranches: the
// wizard needs to trigger this once repo+branch settle and read the
// result imperatively into local state (so it can pre-select, but not
// lock, the build-type tabs), the same "fire on demand, not on mount"
// shape a mutation gives more naturally than a query would here.

import { useMutation } from '@tanstack/react-query'
import { ApiError, readErrorMessage } from '../lib/apiError'

export interface DetectFrameworkInput {
  repoUrl: string
  ref?: string
}

// DetectFrameworkResult mirrors internal/api/build_detect.go's
// detectFrameworkResponse. detected is false whenever frameworkName is
// empty: a repo that can't be cloned or matches no supported provider,
// never surfaced as an error (see that handler's own doc comment), so
// the wizard's fallback is always "keep the manual picker," not a
// scary failure state.
export interface DetectFrameworkResult {
  provider?: string
  frameworkName?: string
  detected: boolean
}

interface RawDetectFrameworkResponse {
  provider?: string
  framework_name?: string
  detected: boolean
}

export async function detectFramework(
  input: DetectFrameworkInput,
): Promise<DetectFrameworkResult> {
  const res = await fetch('/api/v1/build/detect', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      repo_url: input.repoUrl,
      ...(input.ref ? { ref: input.ref } : {}),
    }),
  })
  if (!res.ok) {
    throw new ApiError(
      res.status,
      await readErrorMessage(res, `detect framework failed: ${res.status}`),
    )
  }
  const body = (await res.json()) as RawDetectFrameworkResponse
  return {
    provider: body.provider,
    frameworkName: body.framework_name,
    detected: body.detected,
  }
}

// useDetectFramework is a mutation, not a query: GitBuildSourceFields
// calls .mutate() once a repo URL and branch are both known, the same
// on-demand trigger useGitBranches' own "Load branches" button uses,
// rather than firing automatically on every keystroke.
export function useDetectFramework() {
  return useMutation({
    mutationFn: detectFramework,
  })
}

File: internal/api/build_detect.go
package api

import (
	"context"
	"encoding/json"
	"log/slog"
	"net/http"

	"github.com/GLINCKER/levelrail/internal/build"
)

// detectFunc runs Railpack's own detection (no build, no image) against
// a git source, the seam handleDetectFramework calls through so tests
// can substitute a fake without a real network clone. Defaulted to
// build.Detect in NewRouter.
type detectFunc func(ctx context.Context, req build.DetectRequest) (*build.DetectResult, error)

// detectFrameworkRequest is POST /api/v1/build/detect's body: the same
// repo_url/ref shape gitBranchesRequest and triggerBuildRequest already
// use, so the wizard can call this with the exact values it already has
// once a repo and branch are picked, before any app exists to attach a
// build to.
type detectFrameworkRequest struct {
	RepoURL string `json:"repo_url"`
	Ref     string `json:"ref,omitempty"`
}

// detectFrameworkResponse is POST /api/v1/build/detect's success body.
// Detected is false whenever FrameworkName is empty, spelled out
// explicitly rather than left for the caller to infer from an empty
// string, since "" is also JSON's natural absent-value encoding.
type detectFrameworkResponse struct {
	Provider      string `json:"provider,omitempty"`
	FrameworkName string `json:"framework_name,omitempty"`
	Detected      bool   `json:"detected"`
}

// handleDetectFramework handles POST /api/v1/build/detect: a fast,
// build-free pre-flight check of what Railpack would detect for a git
// source, so the create-app-from-git wizard can show "Detected: Next.js"
// before the operator commits to a build type. Gated at AbilityDeploy,
// the same tier POST /api/v1/git/branches already requires for the same
// "not scoped to an existing app yet" reason (routes.go).
//
// A repo that can't be cloned, or against which Railpack detects
// nothing buildable, is not an error: it responds 200 with
// detected: false, so the wizard falls back to its existing manual
// build-type picker rather than surfacing a scary failure for what is
// often just an unsupported stack.
func (rt *Router) handleDetectFramework(w http.ResponseWriter, r *http.Request) {
	var req detectFrameworkRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}
	if req.RepoURL == "" {
		writeError(w, http.StatusBadRequest, "repo_url is required")
		return
	}

	result, err := rt.detect(r.Context(), build.DetectRequest{RepoURL: req.RepoURL, Ref: req.Ref})
	if err != nil {
		rt.logger.Warn("api: detect framework failed", slog.String("error", err.Error()), slog.String("repo_url", req.RepoURL))
		writeJSON(w, http.StatusOK, detectFrameworkResponse{Detected: false})
		return
	}

	writeJSON(w, http.StatusOK, detectFrameworkResponse{
		Provider:      result.Provider,
		FrameworkName: result.FrameworkName,
		Detected:      result.FrameworkName != "",
	})
}
